City-of-Helsinki/helsinki-paatokset
# PP-367: New endpoint for organization chart data, key_auth authentication method

**To test**
- Checkout branch, run `make new` (take a database dump if necessary)
- Check that the new "Organization" content type exists: https://helsinki-paatokset.docker.so/fi/admin/structure/types/manage/organization/fields
- Add a new user with the "API user" role: https://helsinki-paatokset.docker.so/fi/admin/people/create
  - After creating the user, go to the "Key authentication" tab and make sure an API key was generated
- Create another user with the "Administrator" role:  https://helsinki-paatokset.docker.so/fi/admin/people/create
  - Can't use the root admin for this since it bypasses all access checks
  - Make sure this account has an API key too
- Run the following command to initiate the organization chart import: `drush ahjo-callback:start-org-queue 00001 4 -v`
  - This will only get the first 5 levels, but should be good enough for testing purposes and doesn't take too much time 
- Next run `drush ahjo-callback:list-org-queue` to make sure it was actually added to the queue, then run `drush queue:run ahjo_api_org_queue -v` to actually start the import (this will take a minute or two)
  - This will add more and more items to queue until it reaches the maximum depth specified with the first command
- After the import is done, check the API endpoint: https://helsinki-paatokset.docker.so/fi/ahjo-proxy/org-chart/00001/4
  - The first parameter is the organization ID to start from, and the second level is the maximum depth
- Try opening the same URL in a private browsing window. You should get a 403 error
  - Next add this query parameter to the URL: `?api-key=[API-KEY-FOR-APIUSER]` with the API key from the API user
  - This should work and you should see the JSON output even when you aren't logged in
  - Next, try replacing the query parameter with the API key from the new admin account. You should get a 403 error.
